### PR TITLE
Iteration exporting

### PR DESCRIPTION
**Pull request by IvarStefansson**
## Proposed changes

New tutorial on combining models and exporter. Both standard usage and more exotic iteration exporting.

## Types of changes

What types of changes does your code introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Testing (contribution related mainly to testing of existing functionality)
- [x] Documentation update (if none of the other choices apply)
- [ ] Other: 

## Checklist

_Enter one &check; (`&check;`) in each row. The table can be updated after creating the PR._

||Yes|No|NA|
|---|---|---|---|
|The update is covered by the  test suite |_|_|&check;|
|The documentation is up-to-date |&check;|_|_|
|I have not duplicated existing functionality |&check;|_|_|

### CODE REVIEW

File: tutorials/simulation_exporting.ipynb [hunk 1]
```diff
@@ -0,0 +1,324 @@
+{
```
Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #32.        """Combine exporter class and model using multiple inheritance.
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #34.        Note on inheritance ordering:
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #35.        We want to override the model methods with those
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <p>Suggested change (new text in bold, deletions with strikethrough):</p><p> </p><p>We now extend the <strong>previous example</strong> to the mixed-dimensional case. We can use <code>Exporting</code> as above. In the parameters, we adjust <strong>the</strong> file name to avoid overwriting the previous files. If you inspect the suffixes of the files created, you can see how the exporter deals with multiple time steps by default.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.        """Modify from single- to mixed-dimensional poroelasticity."""
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.            self.mdg, self.box = pp.md_grids_2d.two_intersecting(self.params.get("mesh_args", None),
</pre><p> </p><p>These are the cases where I wish everything to fit into a line. But this is purely esthetics.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #11.                file_name=self.params["file_name"]+"_iterations",
</pre><p> </p><p>The file names are already too long for my taste, could we save some space using <code>self.params["file_name"]+"_iter"</code> instead? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (jhabriel) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #61.            # To make sure the nonlinear iteration index does not interfere with the 
</pre><p> </p><p>I've run the tutorial on my local machine and it works, but I find it quite hard to follow the exporting logic when time and iterations are combined.</p><p> </p><p>If I understand correctly, the output files, say, for 1d domains will be: </p><p> </p><p>debug_iterations_1_000000.pvd  # time = 0 sec</p><p>debug_iterations_1_000101.pvd  # time = 1 sec; iteration = 1</p><p>debug_iterations_1_000102.pvd  # time = 1 sec; iteration = 2</p><p>debug_iterations_1_000201.pvd  # time = 2 sec; iteration = 1</p><p> </p><p>Is that correct? I was wondering if ParaView will be pleased if we add yet another "_" . Something like:</p><p> </p><p>debug_iterations_1_000000_1.pvd  # time = 0 sec</p><p>debug_iterations_1_000001_1.pvd  # time = 1 sec; iteration = 1</p><p>debug_iterations_1_000001_2.pvd  # time = 1 sec; iteration = 2</p><p>debug_iterations_1_000002_1.pvd  # time = 2 sec; iteration = 1</p><p> </p><p>Have you tried something like this?</p><p> </p><p>In any case, it might be beneficial to add a markdown cell below each mixin model with the outputs that it will generate (not all outputs, perhaps just for the bulk or the fracture), with some small in-line comment as included above. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: src/porepy/grids/standard_grids/md_grids_2d.py [hunk 2]
```diff
@@ -19,6 +20,16 @@
 import porepy.grids.standard_grids.utils as utils
 
 
+def _n_cells(mesh_args: Union[np.ndarray, dict, None]):
```
Comment (jhabriel) [inline src/porepy/grids/standard_grids/md_grids_2d.py:23]: Could you please provide documentation for this private method?

File: src/porepy/grids/standard_grids/md_grids_2d.py [hunk 3]
```diff
@@ -19,6 +20,16 @@
...
+    if mesh_args is None:
+        return np.array([2, 2])
+    else:
+        if isinstance(mesh_args, list):
+            mesh_args = np.array(mesh_args)
+        assert type(mesh_args) == np.ndarray
```
Comment (jhabriel) [inline src/porepy/grids/standard_grids/md_grids_2d.py:29]: Isn't more correct to use `isinstance(mesh_args, np.ndarray)` instead?

**Comment by jhabriel** [comment]
Great job @IvarStefansson. 

I think this can be quite useful, specially for debugging complex problems, where from one iteration to another, things can change abruptly. 

I left some comments / suggestions, and in some places I asked for a bit of clarification.

File: tutorials/simulation_exporting.ipynb (hunk 1 above)
Comment (OmarDuran) [reply tutorials/simulation_exporting.ipynb:1]: <p>In agreement with @Jhabriel. It is necessary to explain increasing indices with more elaborate suffixes.</p>

**Comment by OmarDuran** [comment]
<p>This is a nice contribution @IvarStefansson, it is also useful to visualize residuals. I have left only a single suggestion.&nbsp;&nbsp;</p>

**Review by OmarDuran — APPROVED** [review]

File: tutorials/simulation_exporting.ipynb [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
```
Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #15.            matrix = [self._nd_subdomain()]
</pre><p>I see the benefits of using mixin-concepts, but I really don't like this referal to a non-existent attribute. While there should be little room for confusion in this instance, thus it may be acceptable, it goes against all principles of object oriented design. The 'correct' solution is to move the methods to a superclass, or to restructure the code in a different way. Also, we should be cautious to put such code in a tutorial, since they are naturally read as establishing good practice. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.        def _export_data(self):
</pre><p>Just to check: We could also place the original mecthod in AbstractModel and override it locally?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #13.            # FIXME: EK, should we rather use the dof manager for this?
</pre><p>Yes, I believe that would be cleaner.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.            super().prepare_simulation()
</pre><p>Again, I really don't like this - the code makes no sense on its own terms (super has no such method), and is prone to misuse.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #40.                restriction = projection.cell_restriction([sd]).parse(self.mdg)
</pre><p>Again, the DofManager seems like a better option.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: src/porepy/grids/standard_grids/md_grids_2d.py [hunk 5]
```diff
@@ -19,6 +20,17 @@
 import porepy.grids.standard_grids.utils as utils
 
 
+def _n_cells(mesh_args: Union[np.ndarray, dict, None]):
```
Comment (keileg) [inline src/porepy/grids/standard_grids/md_grids_2d.py:23]:
```suggestion
def _n_cells(mesh_args: Union[np.ndarray, dict, None]) -> np.ndarray:
```

File: src/porepy/models/contact_mechanics_biot_model.py [hunk 6]
```diff
@@ -132,6 +133,10 @@ def after_newton_convergence(
         super().after_newton_convergence(solution, errors, iteration_counter)
         self._save_mechanical_bc_values()
 
+    def after_simulation(self):
```
Comment (keileg) [inline src/porepy/models/contact_mechanics_biot_model.py:136]:
```suggestion
    def after_simulation(self) -> None:
```

File: tutorials/simulation_exporting.ipynb [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    "        This method is typically called by _after_newton_convergence.\n",
+    "        \"\"\"\n",
+    "        self.exporter.write_vtu(self._export_data(), time_dependent=True)\n",
+    "        \n",
+    "class ExportingPoroelasticity(Exporting, pp.ContactMechanicsBiot):\n",
+    "    \"\"\"Combine exporter class and model using multiple inheritance.\n",
+    "    \n",
+    "    Note on inheritance ordering:\n",
+    "    We want to override the model methods with those\n",
+    "    implemented in Exporter, which thus should come first. Methods not \n",
```
Comment (jwboth) [inline tutorials/simulation_exporting.ipynb:56]:
```suggestion
    "    implemented in Exporting, which thus should come first. Methods not \n",
```

**Review by jwboth — APPROVED** [review]
Regarding the exporting, I like the demonstration of the advanced and tailored use of it very much. After all, this was what the new format was also about.

Regarding multiple inheritance, I mayeb have not such a strong opinion as Eirik. I am OK with its use, favoring the short length of the code, although I see his point in the use of super.

(I hope I have used ReviewNB correctly - it is the first time)

File: tutorials/simulation_exporting.ipynb (hunk 4 above)
Comment (jwboth) [inline tutorials/simulation_exporting.ipynb:1]: <p>You seem to use the word 'mixin' on purpose - it is also used further up. What does it mean?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (jwboth) [inline tutorials/simulation_exporting.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #50.            data =  displacements + jumps + pressures
</pre><p>Having porepy novices in mind. Although, I honestly very much like your proper demonstration of how to define different data, I wonder whether this last example should not only focus on the iterations. The exporting of displacement jumps has been covered e.g. in TailoredExporting. And there is not structural difference between exporting displacements and pressures. Would it make sense to reduce the exported amount of data in this example to allow the user to grasp the essence?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: src/porepy/viz/exporter.py [hunk 8]
```diff
@@ -1869,7 +1869,7 @@ def _export_polyhedron_3d(self, grids: Iterable[pp.Grid]) -> Meshio_Geom:
                 cf_indices = grid.cell_faces.indices
                 fn_indptr = grid.face_nodes.indptr
                 fn_indices = grid.face_nodes.indices
-
+                # FIXME: Close parenthesis
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/viz/exporter.py:1872]: The left parenthesis in line 1873 is not matched by a right parenthesis

File: tutorials/simulation_exporting.ipynb (hunk 4 above)
Comment (IvarStefansson) [reply tutorials/simulation_exporting.ipynb:1]: Do you mean an abstract version in AbstractModel?

Comment (IvarStefansson) [reply tutorials/simulation_exporting.ipynb:1]: It's a class which doesn't interfere with the "main" class inheritance. But as Eirik points out, I haven't done it properly in this PR

Comment (IvarStefansson) [reply tutorials/simulation_exporting.ipynb:1]: I wholeheartedly agree with you but for two reasons:
1) It may be of interest to show how different data (primary/secondary variables) are fetched from the iterate dictionary.
2) This part of the tutorial is not intended for porepy novices.

I can be persuaded to change this, though.

Comment (IvarStefansson) [reply tutorials/simulation_exporting.ipynb:1]: Hmm. I see what you mean. But according to this [post](https://www.pythontutorial.net/python-oop/python-mixin/), a mixin may call non-implemented methods.

Comment (jwboth) [reply tutorials/simulation_exporting.ipynb:1]: Thanks for the explanation!

Comment (jwboth) [reply tutorials/simulation_exporting.ipynb:1]: I am OK with leaving it.

Comment (keileg) [reply tutorials/simulation_exporting.ipynb:1]: I see the point - let's go with this.

Comment (keileg) [inline tutorials/simulation_exporting.ipynb:1]: <p>I suggest we add a warning regarding Mixin classes - something like the following:</p><p> </p><p>The below code uses a design pattern known as <em>Mixin classes</em>. This is a technique that (in addition to other benefits) allows for sharing of code between classes without using traditional inheritance. While this is a powerful technique, it can also easily lead to confusion, bugs, and code of low quality. Be sure that you read up proper use of Mixin before taking it into use in your own code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/708/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by keileg — APPROVED** [review]

**Review by jhabriel — APPROVED** [review]

File: tutorials/simulation_exporting.ipynb (hunk 4 above)
Comment (jwboth) [reply tutorials/simulation_exporting.ipynb:1]: I am supporting that. (Maybe especially because I have not heard of these classes before.)

Comment (jhabriel) [reply tutorials/simulation_exporting.ipynb:1]: I also support @keileg suggestion.

Comment (IvarStefansson) [reply tutorials/simulation_exporting.ipynb:1]: On second thought, that would require two steps: Extracting u_j and reconstructing jump on fractures. Furthermore, the current solution generalizes to any (compound) ad operator. I'll leave it as is
